src/cluster.c

```diff
@@ -852,176 +689,306 @@ static dict *parse_cluster_slots(valkeyClusterContext *cc, valkeyReply *reply) {
     return NULL;
 }
 
-/**
- * Parse the "cluster nodes" command reply to nodes dict.
- */
-static dict *parse_cluster_nodes(valkeyClusterContext *cc, valkeyReply *reply) {
-    int ret;
-    dict *nodes = NULL;
-    dict *nodes_name = NULL;
-    valkeyClusterNode *master, *slave;
-    cluster_slot *slot;
-    char *pos, *start, *end, *line_start, *line_end;
-    char *role;
-    int role_len;
-    int slot_start, slot_end, slot_ranges_found = 0;
-    sds *part = NULL, *slot_start_end = NULL;
-    int count_part = 0, count_slot_start_end = 0;
-    int k;
-    int len;
-
-    if (reply->type != VALKEY_REPLY_STRING) {
-        valkeyClusterSetError(cc, VALKEY_ERR_OTHER, "Unexpected reply type");
-        goto error;
+/* Store a parsed replica node in given dict using the primary_id as key.
+ * Additional replicas for a primary are stored within the first replica. */
+static int store_replica_node(dict **replicas, char *primary_id, valkeyClusterNode *node) {
+    if (*replicas == NULL) {
+        *replicas = dictCreate(&clusterNodesDictType, NULL);
+        if (replicas == NULL)
+            return VALKEY_ERR;
     }
 
-    nodes = dictCreate(&clusterNodesDictType, NULL);
-    if (nodes == NULL) {
-        goto oom;
+    sds key = sdsnew(primary_id);
+    if (key == NULL)
+        return VALKEY_ERR;
+
+    dictEntry *de = dictFind(*replicas, key);
+    if (de == NULL) {
+        if (dictAdd(*replicas, key, node) != DICT_OK) {
+            sdsfree(key);
+            return VALKEY_ERR;
+        }
+        return VALKEY_OK;
     }
 
-    start = reply->str;
-    end = start + reply->len;
+    /* Store replica node in the existing replica node. */
+    sdsfree(key);
+    valkeyClusterNode *n = dictGetEntryVal(de);
+    if (n->slaves == NULL) {
+        n->slaves = listCreate();
+        if (n->slaves == NULL)
+            return VALKEY_ERR;
+        n->slaves->free = listClusterNodeDestructor;
+    }
+    if (listAddNodeTail(n->slaves, node) == NULL)
+        return VALKEY_ERR;
 
-    line_start = start;
+    return VALKEY_OK;
+}
 
-    for (pos = start; pos < end; pos++) {
-        if (*pos == '\n') {
-            line_end = pos - 1;
-            len = line_end - line_start;
+/* Move parsed replica nodes from the collection to related primary. */
+static int move_replica_nodes(dict *replicas, dict *nodes) {
```

Comment:
From the comment and the names here, it's not entirely clear to me what this does.

"The related primaries" = `nodes`?

Why are they dicts? In the context where they're used, it probably makes sense, but here I'm wondering if the node is the key or the value in each dict...

Reply:
Added comments with some more details.